AmplifyPlugins/API/AWSAPICategoryPlugin/Core/AppSyncList+Paginatable.swift

```diff
@@ -0,0 +1,85 @@
+            return
+        }
+
+        let request = AppSyncList.requestForNextPage(nextToken: nextToken, variables: variables)
+
+        Amplify.API.query(request: request) { result in
+            switch result {
+            case .success(let graphQLResponse):
+                switch graphQLResponse {
+                case .success(let list):
+                    completion(.success(list))
+                case .failure(let graphQLError):
+                    completion(.failure(.listOperation(
+                                            "The AppSync response returned successfully with GraphQL errors.",
+                                            "Check the underlying error for the failed GraphQL response.",
+                                            graphQLError)))
+                }
+            case .failure(let apiError):
+                completion(.failure(.listOperation("The AppSync request failed",
+                                                   "Check the underlying `APIError`",
+                                                   apiError)))
+            }
+        }
+    }
+
+    // This method will always successfully return a new GraphQLRequest with the next token added to the input variables
+    // It leverages the existing GraphQLRequest document builders and decorators to create
+    // - A GraphQL `query` operation
+    // - A `list` type query operation
+    // - The next token added to the GraphQL request variables
+    // If the existing `variables` from current list (`Self`) contains information such as a filter and limit, they are
+    // extracted back out and added back onto the GraphQL request variables to maintain the expected behavior
+    static func requestForNextPage(nextToken: String,
+                                   variables: [String: JSONValue]?) -> GraphQLRequest<AppSyncList<ModelType>> {
+        var documentBuilder = ModelBasedGraphQLDocumentBuilder(modelSchema: ModelType.schema, operationType: .query)
+        documentBuilder.add(decorator: DirectiveNameDecorator(type: .list))
+
+        // Since the original request constructed with object of type `QueryPredicate`, and the type is lost when
+        // translated to a GraphQLRequest. The following extracts the existing filter variables stored in the
+        // GraphQLRequest's variables and uses FilterDecorator to re-create the proper document.
+        if let storedVariables = variables,
+           let filters = storedVariables["filter"],
+           case let .object(filterValue) = filters {
+            let encoder = JSONEncoder()
+            encoder.dateEncodingStrategy = ModelDateFormatting.encodingStrategy
+            guard let filterVariablesData = try? encoder.encode(filterValue),
+                  let filterVariablesJSON = try? JSONSerialization.jsonObject(with: filterVariablesData)
+                    as? [String: Any] else {
+                fatalError("Filter variables is not valid JSON object")
```

Comment:
this is not expected to happen since the original request would have suceeded for us to construct the AppSyncListPayload, thus i believe we should keep it this way. I've also updated it to `assert` instead of `fatalError`. If we return a validation on completion, there isn't anything the developer can do about the serialization failure